Extension/lib/webrequest.js

```diff
@@ -129,40 +129,70 @@
             filterSafebrowsing(tab, requestUrl);
         }
 
-        /*
-         Websocket check.
-         If 'ws://' request is blocked for not existing domain - it's blocked for all domains.
-         Then we gonna limit frame sources to http to block src:'data/text' etc.
-         More details in these issue:
-         https://github.com/AdguardTeam/AdguardBrowserExtension/issues/344
-         https://github.com/AdguardTeam/AdguardBrowserExtension/issues/440
+        if (requestType === adguard.RequestTypes.DOCUMENT || requestType === adguard.RequestTypes.SUBDOCUMENT) {
+            return modifyCSPHeader(requestDetails);
+        }
+    }
 
-         WS connections are detected as "other"  by ABP
-         EasyList already contains some rules for WS connections with $other modifier
-         */
-        var checkWebsocket = (requestType === adguard.RequestTypes.DOCUMENT || requestType === adguard.RequestTypes.SUBDOCUMENT);
+    /**
+     * Modify CSP header to block websocket connections and web workers
+     * @param requestDetails
+     * @returns {{responseHeaders: *}}
+     */
+    function modifyCSPHeader(requestDetails) {
 
-        // Please note, that we do not check WS in Edge:
+        // Please note, that we do not modify response headers in Edge:
         // https://github.com/AdguardTeam/AdguardBrowserExtension/issues/401
         if (adguard.utils.browser.isEdgeBrowser()) {
-            checkWebsocket = false;
+            return;
         }
 
-        if (checkWebsocket) {
-
-            var frameUrl = adguard.frames.getFrameUrl(tab, requestDetails.frameId);
-            var websocketCheckUrl = "ws://adguardwebsocket.check/" + adguard.utils.url.getDomainName(frameUrl);
-            if (adguard.webRequestService.checkWebSocketRequest(tab, websocketCheckUrl, frameUrl)) {
-                var cspHeader = {
-                    name: 'Content-Security-Policy',
-                    value: 'connect-src http: https:; frame-src http: https:; child-src http: https:'
-                };
-                responseHeaders.push(cspHeader);
-                return {
-                    responseHeaders: responseHeaders,
-                    modifiedHeaders: [cspHeader]
-                };
+        var tab = requestDetails.tab;
+        var requestUrl = requestDetails.requestUrl;
+        var responseHeaders = requestDetails.responseHeaders;
+
+        var frameUrl = adguard.frames.getFrameUrl(tab, requestDetails.frameId);
+        var websocketCheckUrl = "ws://adguardwebsocket.check/" + adguard.utils.url.getDomainName(frameUrl);
+        var websocketRule = adguard.webRequestService.getRuleForRequest(tab, websocketCheckUrl, frameUrl, adguard.RequestTypes.WEBSOCKET);
+        var websocketBlocked = adguard.webRequestService.isRequestBlockedByRule(websocketRule);
+
+        var workerRule = adguard.webRequestService.getRuleForRequest(tab, 'blob:', frameUrl, adguard.RequestTypes.SCRIPT);

```

Comment:
Do we really need to look for a rule twice?

Also, I am not so sure now that we should print these requests to the filtering log every time, it'd be a mess.

Let's put it there only in the case if a request was actually blocked.

Also, I'd like us to put something readable instead of the request URL.

For instance:
`content-security-policy-check`